camayoc/tests/qpc/cli/test_reports.py

```diff
@@ -532,6 +534,93 @@ def test_merge_report(merge_by, isolated_filesystem, qpc_server_config):
         assert_json_report_fields(report_item.keys())
 
 
+@pytest.mark.runs_scan
+def test_upload_report(isolated_filesystem, qpc_server_config, tmp_path):
+    """Ensure can upload reports using JSON files.
+
+    :id: fb661698-4cc7-4b6e-9964-a006633eb5af
+    :description: Ensure (JSON) details reports can be uploaded. Also ensure
+        that that uploaded report has been successfully processed by generating
+        a deployments report.
+    :steps:
+        1) Run ``qpc report upload --json-file <json-file>``. Note the
+           job_id and report_id.
+        2) Wait until job completion (``qpc scan job --id <job_id>``).
+        3) Run ``qpc report details --report <report_id>``.
+        4) Ensure details report is equivalent to the uploaded report
+        5) Run ``qpc report deployments --report <job_id>``.
+        6) Ensure the deployments report has the expected number of aggregated
+           reports and each report have the expected fields.
+    :expectedresults: The details and deployments report must be on the requested format
+        (JSON) and must have contents equivalent to the original report.
+    """
+
+    def _get_comparable_source_list(details_report_json):
+        """
+        Return a list of comparable 'sources' from details report.
+
+        In order to be comparable, sources are sorted and "facts" are
+        replaced with a simple count ("facts_count").
+
+        TODO: when details report is finally replaced with something better
+        (and easier to compare), revisit this and compare results at raw fact
+        level.
+        """
+        # preserve the original data
+        source_list = deepcopy(details_report_json["sources"])
+        # given source names are uuids, let's use those to order sources
+        source_list = sorted(source_list, key=itemgetter("source_name"))
+        for source_dict in source_list:
+            source_dict["facts_qty"] = len(source_dict.pop("facts"))
```

Comment:
We are fundamentally comparing that two reports have the same sources, and each source has the same number of facts (+ report_version, server_id and source_type). Could we try to make stronger assertions here? Maybe there are reasons to not compare all the facts, but perhaps at least some of the them that we know are stable? Something similar to what we do in `camayoc/tests/qpc/api/v1/reports/test_reports.py`.

Reply:
> Something similar to what we do in camayoc/tests/qpc/api/v1/reports/test_reports.py.

Do you mean a specific test?

> Could we try to make stronger assertions here?

The problem is that facts is a list of dicts that IDK how to properly sort to compare.
I could learn how to by studying the raw facts for each source type, but I'm not sure if it is worth the effort (considering we will move away from it soon ™️ ).

Reply:
Regardless, I will compare the generated fingerprints and the original ones because they are based on a "proper table" (instead of a bunch of JSON Fields). Let's revisit this once I'm done.

Reply:
> Do you mean a specific test?

Basically any of them :) . They all follow the same pattern - create two large dicts and then iterate through one, while comparing data to another.

If sorting is the only problem, then we could switch json facts from list to set, or use [deepdiff](https://pypi.org/project/deepdiff/) that can deal with sorting automatically (that would require adding new dependency, as somehow deepdiff never found a way into Camayoc).

But let's continue once you are done with the work.